core/dbt/task/printer.py

```diff
@@ -98,10 +98,10 @@ def print_run_result_error(result, newline: bool = True, is_warning: bool = Fals
                 )
             )
 
-        if result.message:
-            fire_event(RunResultError(msg=result.message))
-        else:
-            fire_event(RunResultErrorNoMessage(status=result.status))
+            if result.message:
```

Comment:
@jtcohen6 @nathaniel-may Does this fix make sense in terms of desired behavior?  The logic isn't clear or documented about when things have messages and when things are warnings and what the intersection of the two should be so this is my ian-makes-assumptions solution.

Reply:
We are setting `result.message` when a test warns:
https://github.com/dbt-labs/dbt-core/blob/5e0a7659174a9277e2013390f8615acc97464d40/core/dbt/task/test.py#L194-L196

So the expectation is that we should keep firing that event, and printing it to the CLI:
```
08:14:33  1 of 1 START test unique_my_model_id............................................ [RUN]
08:14:33  1 of 1 WARN 1 unique_my_model_id................................................ [WARN 1 in 0.04s]
08:14:33
08:14:33  Finished running 1 test in 0.17s.
08:14:33
08:14:33  Completed with 1 warning:
08:14:33
08:14:33  Warning in test unique_my_model_id (models/tests.yml)
08:14:33    Got 1 result, configured to warn if != 0     <<< removed by the change in this PR
08:14:33
08:14:33    compiled SQL at target/compiled/testy/models/tests.yml/unique_my_model_id.sql
```

The difference is, that message should have a `WARN` log level, instead of an `ERROR` level.

**Simpler fix:** I imagine that might look like adding a new event type, and rearranging the logic here, so that `if is_warning and result.message`, we fire `RunResultWarningMessage`. (And rename event type Z024 from `RunResultError` to `RunResultErrorMessage`.)

**Better fix:** To turn these into real structured events, it sure seems like we should be passing in `num_errors` + `test.config.error_if`/`test.config.warn_if` as dataclass attributes, and synthesizing the warning/error messages in the events module, rather than constructing those strings in `task/test.py`.

Reply:
Just spoke with @iknox-fa over zoom. This task was not created as a part of the structured logging effort, and is in dire need of a refactor. It also clearly calls out for the event module to handle "error on warning" behavior so I have made a tech debt ticket for it: #4914.